src/gui/core/canvas.cpp

```diff
@@ -844,33 +834,27 @@ void canvas::draw(const bool force)
 		variables_.add("height", wfl::variant(h_));
 	}
 
+	auto renderer = std::unique_ptr<SDL_Renderer, decltype(&SDL_DestroyRenderer)> {nullptr, &SDL_DestroyRenderer};
```

Comment:
Is it possible to simplify this by leveraging template deduction guides? If `std::unique_ptr` has these properly configured then I would expect the following to work:

```
auto renderer = std::unique_ptr(SDL_CreateSoftwareRenderer(canvas_), &SDL_DestroyRenderer);
```

Or maybe even:

```
std::unique_ptr renderer(SDL_CreateSoftwareRenderer(canvas_), &SDL_DestroyRenderer);
```

On a related note, there doesn't seem to be any reason to create the renderer in the if statement, as both branches create it in the same way.

Reply:
Depending on which branch we're going to take, `canvas_` might be null at this point. So to move the `SDL_CreateSoftwareRenderer` call out of the `if`, the code would end looking as if it had 4 code paths even though it's only 2:

```
const bool must_clear_canvas = canvas_;
if(!canvas_) {
    DBG_GUI_D << "Canvas: create new empty canvas.\n";
    canvas_ = surface(w_, h_);
}
std::unique_ptr renderer(SDL_CreateSoftwareRenderer(canvas_), &SDL_DestroyRenderer);
if(must_clear_canvas) {
    DBG_GUI_D << "Canvas: clearing cached canvas.\n";
    SDL_SetRenderDrawColor(renderer.get(), 0, 0, 0, 0);
    SDL_RenderClear(renderer.get());
}
```

Reply:
Hmm… I can't decide whether that's okay or not…

Even if it's not, there's also the possibility of casting nullptr:

```
std::unique_ptr renderer(static_cast<SDL_Renderer*>(nullptr), &SDL_DestroyRenderer);
```

Alternatively, don't rely on deduction guides but also don't use auto:

```
std::unique_ptr<SDL_Renderer> renderer(nullptr, &SDL_DestroyRenderer);
```

Reply:
GCC 10 refuses to compile either of those, are you sure they're valid?

Reply:
I thought they would be…? Do the two from my other post work? I don't think there's any semantic difference between the second one in my first post and the first in my second post. Unless it doesn't like casting nullptr to a specific pointer type, but that seems unlikely.

Reply:
https://stackoverflow.com/questions/51109767/why-cant-unique-ptrs-template-arguments-be-deduced

Reply:
So the answer is pretty much no. `std::make_unique` might help, but then you can't pass `SDL_DestroyRenderer`.

Reply:
I wondered `auto renderer = create_auto_deleter<SDL_DestroyRenderer>();` was possible, with the template deducing the type `SDL_Renderer` from deleter's arguments. I think there is, but didn't reach a working template before deciding that the current code is going to be easier to read.

Reply:
So in other words, none of my examples work, huh? That's unfortunate…